### ISSUE TITLE
piwik and CRS v3

### ISSUE THREAD

**Issue by quenenni** [SpiderLabs/owasp-modsecurity-crs#388]
First.. I just wanted to say that the v3 is very very nice. You did a hell of a good job. Thanks a lot for making our life less paranoiac :+1: 

We are offering / using Piwik instead of google-analytics on our server.
But the way Piwik is using Cookies makes all the requests blocked by modsec.

Here is an example of Piwik Cookies:
```
Cookie: _pk_id.3.7967=bba7d2c3d4e7d2ab.1460702900.22.1467274857.1467274839.; _pk_ref.3.7967=%5B%22%22%2C%22%22%2C1467274839%2C%22https%3A%2F%2Fwww.google.be%22%5D; spip_lang=fr; spip_lang_ecrire=fr; _pk_ses.3.7967=*
```

The Cookie _pk_ref.3.7967 with the value %5B%22%22%2C%22%22%2C1467274839%2C%22https%3A%2F%2Fwww.google.be%22%5D (_pk_ref.3.7967=["","",1467274839,"https://www.google.be"]) generates the problems with the rules:

- 942370: Detects classic SQL injection probings 2/2
- 942200: Detects MySQL comment-/space-obfuscated injections and backtick termination

I resolved it by creating a custom rule:
```
SecRule &REQUEST_COOKIES_NAMES:/^_pk_ref/ "@eq 1"               "phase:1,id:13000,nolog,auditlog,chain,pass,\
        ctl:ruleRemoveById=942370,\
        ctl:ruleRemoveById=942200"
  SecRule &REQUEST_COOKIES_NAMES:/^_pk_id/ "@eq 1"              "chain"
  SecRule &REQUEST_COOKIES_NAMES:/^_pk_ses/ "@eq 1"
```

Not sure it's the best way as anyone can manipulate Cookies as they want, but it's working.

I just wanted you to be aware of this.

**Comment by csanders-git** [comment]
Thank you for the report. I will say that this is not the best way to add exceptions as it occurs on a transaction by transaction basis. Instead you can update the rules at run time to add that exception by using 'SecRuleUpdateTargetById' (https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual#SecRuleUpdateTargetById)

Generally it'd look like SecRuleUpdateTargetByID 942370 "!REQUEST_COOKIES_NAMES:_pk_ref"

**Comment by quenenni** [comment]
Excellent advice.
Thanks

**Comment by dune73** [comment]
The two rules in question are PL2. So I think it is OK to expect the user to be able to tune as explained above.
 
Still, piwik is a generic and widspread piece of software with a bad taste for cookie design (it's not the only taste problem with piwik. In fact the architecture forces you to expose the admin interface together with the page counter. If it would be seperate, the admin would be far easier to protect.)

So would it might make some sense to add SecRuleUpdateTargetByID statements right into the CRS for cases like this.

@quenenni : Your tuning attempt removes the said rules completely; also for ARGS.
@csanders-git : I think you are not targeting the cookie but its name. I guess this should be `SecRuleUpdateTargetByID 942370 "!REQUEST_COOKIES:/^_pk_/` (not tested)

**Comment by quenenni** [comment]
@dune73 : I know. I generally resolve blocked requests by doing a quick removal of the rule (the most precisely I can, but for this one, I had nothing to work on.. these Cookies where on every types of requests).
Then I try to make a better rule.

So, I tried @csanders-git idea and here what's working:
```
SecRuleUpdateTargetByID 942370 "!REQUEST_COOKIES_NAMES:/^_pk_ref/"
SecRuleUpdateTargetByID 942200 "!REQUEST_COOKIES:/_pk_ref/"
```

What I don't understand is why it doesn't work when I try this (for the second line):
```
SecRuleUpdateTargetByID 942200 "!REQUEST_COOKIES_NAMES:/^_pk_ref/"
```
'_pk_ref' is a Cookie name, and it works for the rule 942370. So why not for 942200.. 
```
Cookie: _pk_id.3.7967=bbce3045b3017cc3.1449570374.25.1467369104.1467368235.; _pk_ref.3.7967=%5B%22%22%2C%22%22%2C1467368235%2C%22http%3A%2F%2Fwww.oxfammagasinsdumonde.be%2F2015%2F12%2Fet-si-cette-annee-on-inventait-un-nouvel-arbre-de-noel%2F%22%5D; _pk_ses.3.7967=*
```

Well, now, it's working well.

**Comment by quenenni** [comment]
> Still, piwik is a generic and widspread piece of software with a bad taste for cookie design (it's not the only taste problem with piwik

Every CMS I have to manage has plenty of problems like this one.
As it's a new server with the new v3 rules, I decided to start over with all the exceptions I will encounter.

One thing that was incredible with the test I did is the CMS Drupal.
Amazing how everything was good, no custom rules to create after having test all I could in the admin section of the website.
Congrats to them. They seem to take security more seriously than the others.
